File: frontend/src/types/support.ts
export interface SupportMessage {
  uuid: string
  user_id: string
  user_name: string | null
  content: string
  is_support_reply: boolean
  created_at: string | null
  edited_at: string | null
}

export interface SupportAttachment {
  uuid: string
  filename: string
  file_type: string | null
  uploaded_by: string
  message_uuid: string | null
  created_at: string | null
}

export interface SupportWatcher {
  user_id: string
  name: string
  email: string | null
}

export interface SupportTicket {
  uuid: string
  // Human-friendly sequential id (e.g. 1024). Nullable for any legacy ticket
  // created before this feature shipped that hasn't been backfilled yet.
  ticket_number: number | null
  subject: string
  status: 'open' | 'in_progress' | 'closed'
  priority: 'low' | 'normal' | 'high'
  user_id: string
  user_name: string | null
  user_email: string | null
  team_id: string | null
  assigned_to: string | null
  category: string | null
  // Only present for support agents — backend strips this for ticket owners.
  tags?: string[]
  // Users tagged to follow the ticket. Visible to anyone who can see it.
  watchers: SupportWatcher[]
  messages: SupportMessage[]
  attachments: SupportAttachment[]
  message_count: number
  created_at: string | null
  updated_at: string | null
  closed_at: string | null
}

export interface SupportTicketSummary {
  uuid: string
  ticket_number: number | null
  subject: string
  status: 'open' | 'in_progress' | 'closed'
  priority: 'low' | 'normal' | 'high'
  user_id: string
  user_name: string | null
  assigned_to: string | null
  category: string | null
  // Only present for support agents — backend strips this for ticket owners.
  tags?: string[]
  // user_ids of watchers; used to render a "Watching" badge without a detail fetch.
  watcher_ids: string[]
  message_count: number
  last_message_preview: string | null
  last_message_at: string | null
  last_message_is_support_reply: boolean | null
  last_message_user_id: string | null
  read_by: string[]
  created_at: string | null
  updated_at: string | null
  closed_at: string | null
}

export interface SupportContact {
  user_id: string
  email: string
  name: string
}

File: backend/app/models/support.py
"""Support ticket models."""

import datetime
import uuid as uuid_mod
from enum import Enum
from typing import Optional

from beanie import Document
from pydantic import BaseModel, Field


class TicketStatus(str, Enum):
    OPEN = "open"
    IN_PROGRESS = "in_progress"
    CLOSED = "closed"


class TicketPriority(str, Enum):
    LOW = "low"
    NORMAL = "normal"
    HIGH = "high"


class SupportMessage(BaseModel):
    """Embedded message within a ticket conversation."""

    uuid: str = Field(default_factory=lambda: uuid_mod.uuid4().hex)
    user_id: str
    user_name: Optional[str] = None
    content: str
    is_support_reply: bool = False
    # Internal notes are written by support agents to coordinate with each
    # other on a ticket. They are never returned to the ticket owner or to
    # non-support watchers, and don't notify or email the requester.
    is_internal_note: bool = False
    created_at: datetime.datetime = Field(
        default_factory=lambda: datetime.datetime.now(datetime.timezone.utc)
    )
    edited_at: Optional[datetime.datetime] = None


class SupportAttachment(BaseModel):
    """File attached to a support ticket."""

    uuid: str = Field(default_factory=lambda: uuid_mod.uuid4().hex)
    filename: str
    file_type: Optional[str] = None
    file_data: str = ""  # base64 encoded (legacy, prefer file_path)
    file_path: Optional[str] = None  # on-disk path (preferred)
    uploaded_by: str
    message_uuid: Optional[str] = None  # linked to a specific message
    created_at: datetime.datetime = Field(
        default_factory=lambda: datetime.datetime.now(datetime.timezone.utc)
    )


class SupportTicket(Document):
    uuid: str = Field(default_factory=lambda: uuid_mod.uuid4().hex)
    # Human-friendly sequential id (e.g. 1024). Assigned at insert time via an
    # atomic counter; legacy tickets created before this feature get backfilled
    # on first read. Optional only so older docs deserialize cleanly.
    ticket_number: Optional[int] = None
    subject: str
    status: TicketStatus = TicketStatus.OPEN
    priority: TicketPriority = TicketPriority.NORMAL

    # Creator
    user_id: str
    user_name: Optional[str] = None
    user_email: Optional[str] = None
    team_id: Optional[str] = None

    # Conversation
    messages: list[SupportMessage] = []
    attachments: list[SupportAttachment] = []

    # Category (None for regular tickets, "feedback_prompt" for check-in prompts)
    category: Optional[str] = None

    # Assignment
    assigned_to: Optional[str] = None  # user_id of support person

    # Read tracking — user_ids who have viewed the ticket since the last message
    read_by: list[str] = []

    # Tags applied by support agents — internal-only; never returned to the
    # ticket owner or other non-support users.
    tags: list[str] = []

    # Users (by user_id) the requester or an agent has looped in on this
    # ticket. Watchers can view the ticket, get notified on new messages
    # and status changes, and reply to it. Distinct from `tags` (which are
    # support-internal labels) — watchers are always visible to everyone
    # who can see the ticket.
    watchers: list[str] = []

    # Timestamps
    created_at: datetime.datetime = Field(
        default_factory=lambda: datetime.datetime.now(datetime.timezone.utc)
    )
    updated_at: datetime.datetime = Field(
        default_factory=lambda: datetime.datetime.now(datetime.timezone.utc)
    )
    closed_at: Optional[datetime.datetime] = None

    class Settings:
        name = "support_ticket"
        indexes = [
            "uuid",
            "ticket_number",
            "user_id",
            "status",
            "assigned_to",
            "tags",
            "watchers",
            [("status", 1), ("created_at", -1)],
            [("user_id", 1), ("created_at", -1)],
        ]


class SupportCounter(Document):
    """Atomic counter for sequential ticket numbers.

    One singleton document keyed by `name` (always "support_ticket"). Mongo's
    findOneAndUpdate with $inc + upsert guarantees each create_ticket call
    gets a unique increasing value, even under concurrent inserts.
    """

    name: str
    value: int = 0

    class Settings:
        name = "support_counter"
        indexes = ["name"]
